src/rustzcash.rs

```diff
@@ -1563,3 +1568,86 @@ pub extern "system" fn librustzcash_sapling_proving_ctx_init() -> *mut SaplingPr
 pub extern "system" fn librustzcash_sapling_proving_ctx_free(ctx: *mut SaplingProvingContext) {
     drop(unsafe { Box::from_raw(ctx) });
 }
+
+#[no_mangle]
+pub extern "system" fn librustzcash_zip32_xsk_master(
+    seed: *const c_uchar,
+    seedlen: size_t,
+    xsk_master: *mut [c_uchar; 169],
+) {
+    let seed = unsafe { std::slice::from_raw_parts(seed, seedlen) };
+
+    let xsk = zip32::ExtendedSpendingKey::master(seed);
+
+    xsk.write(&mut (unsafe { &mut *xsk_master })[..])
+        .expect("should be able to serialize an ExtendedSpendingKey");
+}
+
+#[no_mangle]
+pub extern "system" fn librustzcash_zip32_xsk_derive(
+    xsk_parent: *const [c_uchar; 169],
+    i: uint32_t,
+    xsk_i: *mut [c_uchar; 169],
+) {
+    let xsk_parent = zip32::ExtendedSpendingKey::read(&unsafe { *xsk_parent }[..])
+        .expect("valid ExtendedSpendingKey");
+    let i = zip32::ChildIndex::from_index(i);
+
+    let xsk = xsk_parent.derive_child(i);
+
+    xsk.write(&mut (unsafe { &mut *xsk_i })[..])
+        .expect("should be able to serialize an ExtendedSpendingKey");
+}
+
+#[no_mangle]
+pub extern "system" fn librustzcash_zip32_xfvk_derive(
+    xfvk_parent: *const [c_uchar; 169],
+    i: uint32_t,
+    xfvk_i: *mut [c_uchar; 169],
+) -> bool {
+    let xfvk_parent = zip32::ExtendedFullViewingKey::read(&unsafe { *xfvk_parent }[..])
+        .expect("valid ExtendedFullViewingKey");
+    let i = zip32::ChildIndex::from_index(i);
+
+    let xfvk = match xfvk_parent.derive_child(i) {
+        Ok(xfvk) => xfvk,
+        Err(_) => return false,
+    };
+
+    xfvk.write(&mut (unsafe { &mut *xfvk_i })[..])
+        .expect("should be able to serialize an ExtendedFullViewingKey");
+
+    true
+}
+
+#[no_mangle]
+pub extern "system" fn librustzcash_zip32_xfvk_address(
+    xfvk: *const [c_uchar; 169],
```

Comment:
why is xfvk 169 bytes? I couldn't reach this number:) Would be good to specify this length and explain it in zip 32 when describing encoding of keys

Reply:
It's specified in the `Key Encodings` section of ZIP 32. 1-byte depth, 4-byte tag, 4-byte child index, and then 5 32-byte values. I agree we could improve its documentation here, but we should do that as part of the Rust code refactor, so we have consistent definitions of constants.

Reply:
The diversifier is 32 bits? Not 11?

Reply:
I don't see an explicit length mentioned in the key encodings section.  https://github.com/zcash/zips/blob/813a8891d1374099d592bd5e2c3d29037460f4a7/zip-0032.rst.
It seems implies the diversifer e.g. is 11 bytes

Reply:
The diversifier key (which we use to derive diversifiers) is 32 bytes.

Reply:
Ohh right I mixed it up